lib/ace/ext/papeeria/spellchecker.coffee

```diff
@@ -1,39 +1,235 @@
+
+      return
+
+    # Show popup in editor.
+    # @param {Array} options: list of corrections for the current word.
+    show: (options) ->
+      PopupManager.detach() if editor?
+
+      @options = options
+      @popup.setData(options)
+
+      @activated = true
+
+      position = @editor.getCursorPosition()
+      @base = @session.doc.createAnchor(position.row, position.column)
+
+      @editor.keyBinding.addKeyboardHandler(@keyboard)
+      @editor.on("change", @onChange)
+      @editor.on("changeSelection", @onChangeSelection)
+      @editor.on("blur", @onBlur)
+      @editor.on("mousedown", @onMouseDown)
+      @editor.on("mousewheel", @onMouseWheel)
+      @editor.$blockScrolling = Infinity
+
+      renderer = @editor.renderer
+      lineHeight = renderer.layerConfig.lineHeight
+      pos = renderer.$cursorLayer.getPixelPosition(this.base, true)
+      pos.left -= @popup.getTextLeftOffset()
+      rect = @editor.container.getBoundingClientRect()
+      pos.top += rect.top - renderer.layerConfig.offset
+      pos.left += rect.left - @editor.renderer.scrollLeft
+      pos.left += renderer.gutterWidth
+
+      @popup.show(pos, lineHeight)
+      return
+
+    # Detach popup from editor.
+    detach: ->
+      if @activated
+        @popup.hide()
+        @base.detach()
+
+      @editor.keyBinding.removeKeyboardHandler(@keyboard)
+      @editor.off("change", @onChange)
+      @editor.off("changeSelection", @onChangeSelection)
+      @editor.off("blur", @onBlur)
+      @editor.off("mousedown", @onMouseDown)
+      @editor.off("mousewheel", @onMouseWheel)
+
+      @activated = false
+      return
+
+    # Insert a selected correction (option) instead of the current word.
+    insertCorrection: ->
+      PopupManager.detach()
+      correction = PopupManager.getSelectedCorrection()
+      wordRange = getCurrentWordRange(PopupManager.editor)
+      PopupManager.editor.getSession().replace(wordRange, correction)
+      return
+
+    # Change current row in a popup when "up" or "down" listeners triggers.
+    # @param {String} where: direction.
+    goTo: (where) ->
+      row = @popup.getRow()
+      max = @popup.getSession().getLength() - 1
+
+      switch where
+        when "up"
+          if row <= 0 then row = max else row = row - 1
+        when "down"
+          if row >= max then row = -1 else row = row + 1
+        when "start" then row = 0
+        when "end" then row = max
+
+      @popup.setRow(row)
+      return
+
+    # Pick the chosen option from options object.
+    # @return {String}: chosen correction.
+    getSelectedCorrection: ->
+      row = PopupManager.popup.getRow()
+      return PopupManager.options[row].value
+
+    onChange: (e) ->
+      PopupManager.detach()
+      return
+
+    onChangeSelection: (e) ->
+      if not @activated
+        PopupManager.detach()
+        return
+      cursor = PopupManager.editor.selection.lead
+      if cursor.row isnt @base.row or cursor.column < @base.column
```

Comment:
I bet `@base` is going to be undefined here, because this is not `PopupManager`. Why don't you replace `on...` methods with closures when you add them to events? 

```
editor.on("change", -> PopupManager.onChange())
```

This way you can always be sure that in PopupManager methods `this` will be `PopupManager` itself